### PR TITLE
[enhance](mtmv)Refuse to execute insert overwrite on the same table

### PR DESCRIPTION
**Pull request by zddr**
Refuse to execute insert overwrite on the same table:
- Sometimes it is not possible to cancel a running insert overwrite
- When executing insert overwrite in parallel, it causes temporary partitions to affect each other

doc: https://github.com/apache/doris-website/pull/1139

### CODE REVIEW

**Comment by zddr** [comment]
run buildall

**Comment by zddr** [comment]
run buildall

**Comment by zddr** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/insert/InsertOverwriteTableCommand.java [hunk 1]
```diff
@@ -157,35 +160,90 @@ public void run(ConnectContext ctx, StmtExecutor executor) throws Exception {
...
                 // partitions and return. for transactional, the replacement will really occur when insert successed,
                 // i.e. `insertInto` finished. then we call taskGroupSuccess to make replacement.
                 insertInto(ctx, executor, taskId);
-                Env.getCurrentEnv().getInsertOverwriteManager().taskGroupSuccess(taskId, (OlapTable) targetTable);
+                insertOverwriteManager.taskGroupSuccess(taskId, (OlapTable) targetTable);
             } else {
                 List<String> tempPartitionNames = InsertOverwriteUtil.generateTempPartitionNames(partitionNames);
-                taskId = Env.getCurrentEnv().getInsertOverwriteManager()
+                if (isCancelled) {
+                    LOG.info("insert overwrite isCancelled before registerTask");
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/insert/InsertOverwriteTableCommand.java:179]: it is better print query id or other id to identifier queries

File: fe/fe-core/src/main/java/org/apache/doris/job/extensions/mtmv/MTMVTask.java [hunk 2]
```diff
@@ -253,9 +258,18 @@ public synchronized void onSuccess() throws JobException {
     @Override
     protected synchronized void executeCancelLogic() {
         LOG.info("mtmv task cancel, taskId: {}", super.getTaskId());
+        // If the be scheduling has not been triggered yet, cancel the scheduling first
+        if (command != null) {
+            command.cancel();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/job/extensions/mtmv/MTMVTask.java:263]: could we do insert overwrite cancel in executor.cancel(). and wait real cannel in insert overwrite cancel? current impl is very hard to use

**Comment by zddr** [comment]
run buildall

**Comment by zddr** [comment]
run buildall

**Review by morrySnow — DISMISSED** [review]

**Comment by zddr** [comment]
run buildall

**Comment by zddr** [comment]
run buildall

**Comment by zddr** [comment]
run buildall

**Comment by zddr** [comment]
run performance

**Comment by zddr** [comment]
run buildall

**Comment by zddr** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/insertoverwrite/InsertOverwriteManager.java [hunk 3]
```diff
@@ -270,6 +279,50 @@ private boolean rollback(long taskId) {
...
+     *
+     * @param dbId Run the dbId for insert overwrite
+     * @param tableId Run the tableId for insert overwrite
+     */
+    public void recordRunningTableOrException(long dbId, long tableId) {
+        runningLock.writeLock().lock();
+        try {
+            if (runningTables.containsKey(dbId) && runningTables.get(dbId).contains(tableId)) {
+                throw new AnalysisException(
+                        String.format("insert overwrite is running on db: %s, table: %s", dbId, tableId));
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/insertoverwrite/InsertOverwriteManager.java:294]:
```suggestion
                        String.format("Not allowed running Insert Overwrite on same table: %s.%s", dbId, tableId));
                       
```

And I suggest to use db/table name in error msg

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/insert/InsertOverwriteTableCommand.java [hunk 4]
```diff
@@ -157,35 +160,103 @@ public void run(ConnectContext ctx, StmtExecutor executor) throws Exception {
...
                 // partitions and return. for transactional, the replacement will really occur when insert successed,
                 // i.e. `insertInto` finished. then we call taskGroupSuccess to make replacement.
                 insertInto(ctx, executor, taskId);
-                Env.getCurrentEnv().getInsertOverwriteManager().taskGroupSuccess(taskId, (OlapTable) targetTable);
+                insertOverwriteManager.taskGroupSuccess(taskId, (OlapTable) targetTable);
             } else {
                 List<String> tempPartitionNames = InsertOverwriteUtil.generateTempPartitionNames(partitionNames);
-                taskId = Env.getCurrentEnv().getInsertOverwriteManager()
+                if (isCancelled) {
+                    LOG.info("insert overwrite isCancelled before registerTask, queryId: {}", ctx.getQueryIdentifier());
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/insert/InsertOverwriteTableCommand.java:179]:
```suggestion
                    LOG.info("insert overwrite is cancelled before registerTask, queryId: {}", ctx.getQueryIdentifier());
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/insert/InsertOverwriteTableCommand.java [hunk 5]
```diff
@@ -157,35 +160,103 @@ public void run(ConnectContext ctx, StmtExecutor executor) throws Exception {
...
+                    return;
+                }
+                taskId = insertOverwriteManager
                         .registerTask(targetTable.getDatabase().getId(), targetTable.getId(), tempPartitionNames);
+                if (isCancelled) {
+                    LOG.info("insert overwrite isCancelled before addTempPartitions, queryId: {}",
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/insert/InsertOverwriteTableCommand.java:185]:
```suggestion
                    LOG.info("insert overwrite is cancelled before addTempPartitions, queryId: {}",
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/insert/InsertOverwriteTableCommand.java [hunk 6]
```diff
@@ -157,35 +160,103 @@ public void run(ConnectContext ctx, StmtExecutor executor) throws Exception {
...
+                            ctx.getQueryIdentifier());
+                    // not need deal temp partition
+                    insertOverwriteManager.taskSuccess(taskId);
+                    return;
+                }
                 InsertOverwriteUtil.addTempPartitions(targetTable, partitionNames, tempPartitionNames);
+                if (isCancelled) {
+                    LOG.info("insert overwrite isCancelled before insertInto, queryId: {}", ctx.getQueryIdentifier());
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/insert/InsertOverwriteTableCommand.java:193]: ditto

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/insert/InsertOverwriteTableCommand.java [hunk 7]
```diff
@@ -157,35 +160,103 @@ public void run(ConnectContext ctx, StmtExecutor executor) throws Exception {
...
+                    insertOverwriteManager.taskFail(taskId);
+                    return;
+                }
                 insertInto(ctx, executor, tempPartitionNames);
+                if (isCancelled) {
+                    LOG.info("insert overwrite isCancelled before replacePartition, queryId: {}",
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/insert/InsertOverwriteTableCommand.java:199]: ditto

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/insert/InsertOverwriteTableCommand.java [hunk 8]
```diff
@@ -157,35 +160,103 @@ public void run(ConnectContext ctx, StmtExecutor executor) throws Exception {
...
+                            ctx.getQueryIdentifier());
+                    insertOverwriteManager.taskFail(taskId);
+                    return;
+                }
                 InsertOverwriteUtil.replacePartition(targetTable, partitionNames, tempPartitionNames);
-                Env.getCurrentEnv().getInsertOverwriteManager().taskSuccess(taskId);
+                if (isCancelled) {
+                    LOG.info("insert overwrite isCancelled before taskSuccess, do nothing, queryId: {}",
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/insert/InsertOverwriteTableCommand.java:206]: ditto

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/insert/InsertOverwriteTableCommand.java [hunk 9]
```diff
@@ -157,35 +160,103 @@ public void run(ConnectContext ctx, StmtExecutor executor) throws Exception {
...
+        this.isCancelled = true;
+    }
+
+    /**
+     * wait insert overwrite not running
+     */
+    public void waitNotRunning() {
+        long waitMaxTimeMills = 10 * 1000L;
+        long waitTime = 0;
+        while (true) {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/insert/InsertOverwriteTableCommand.java:240]: You can use `Awaitility`, just like using it in regression test case

File: regression-test/framework/src/main/groovy/org/apache/doris/regression/suite/Suite.groovy [hunk 10]
```diff
@@ -1275,7 +1275,29 @@ class Suite implements GroovyInterceptable {
...
+        Assert.assertEquals("SUCCESS", status)
+    }
+
+    void waitingMTMVTaskFinishedByMvNameAllowCancel(String mvName) {
+        Thread.sleep(2000);
+        String showTasks = "select TaskId,JobId,JobName,MvId,Status,MvName,MvDatabaseName,ErrorMsg from tasks('type'='mv') where MvName = '${mvName}' order by CreateTime ASC"
+        String status = "NULL"
+        List<List<Object>> result
+        long startTime = System.currentTimeMillis()
+        long timeoutTimestamp = startTime + 5 * 60 * 1000 // 5 min
```
Comment (morningman) [inline regression-test/framework/src/main/groovy/org/apache/doris/regression/suite/Suite.groovy:1291]: Use `Awaitility`

**Review by morningman — COMMENTED** [review]
Need to add negative test cases, eg, insert same table at same time, one should fail

File: regression-test/framework/src/main/groovy/org/apache/doris/regression/suite/Suite.groovy (hunk 10 above)
Comment (zddr) [reply regression-test/framework/src/main/groovy/org/apache/doris/regression/suite/Suite.groovy:1291]: There are also some similar methods that I can modify uniformly in other PRs

**Comment by zddr** [comment]
run buildall

**Comment by zddr** [comment]
run buildall

**Review by morningman — APPROVED** [review]

**Review by morrySnow — APPROVED** [review]

**Pull request merged**